File: ui/src/main/scala/tripaint/view/gui/AskForFileOpenSettingsDialog.scala
package tripaint.view.gui

import tripaint.Color
import tripaint.coords.{GridCoords, StorageCoords}
import tripaint.grid.{GridCell, ImageGrid}
import tripaint.image.{ImageStorage, RegularImage}
import tripaint.image.format.StorageFormat
import tripaint.view.FileOpenSettings
import tripaint.view.gui.DialogUtils.{getValueFromCustomDialog, makeGridPane}
import tripaint.view.image.TriImageForPreview

import javafx.collections.FXCollections
import javafx.embed.swing.SwingFXUtils
import javafx.geometry.Orientation
import javafx.scene.control.{ChoiceBox, Label, Separator}
import javafx.scene.control.ButtonType
import javafx.scene.image.ImageView
import javafx.scene.layout.Pane
import javafx.util.StringConverter

import java.io.File
import scala.util.{Success, Try}

object AskForFileOpenSettingsDialog {
  def askForFileOpenSettings(
      imagePreview: (File, Int, Int, Int),
      formats: Seq[(StorageFormat, String)],
      initiallySelectedFormat: Int,
      readImage: File => Option[RegularImage]
  ): Option[FileOpenSettings] = {
    val (previewFile, imageSize, xCount, yCount) = imagePreview
    val (previewWidth, previewHeight) = (xCount * imageSize, yCount * imageSize)

    val xCoordTF = RestrictedTextField.uintTF
    xCoordTF.setPromptText("0")

    val yCoordTF = RestrictedTextField.uintTF
    yCoordTF.setPromptText("0")

    val formatMap: Map[StorageFormat, String] = Map.from(formats)

    val formatChooser = {
      val b = new ChoiceBox[StorageFormat](FXCollections.observableArrayList(formats.map(_._1)*))
      b.getSelectionModel.select(initiallySelectedFormat)
      b.setConverter(new StringConverter[StorageFormat] {
        override def toString(t: StorageFormat) = formatMap(t)
        override def fromString(s: String): StorageFormat =
          throw new UnsupportedOperationException()
      })
      b
    }

    val resultFromInputs = () => {
      val xt = xCoordTF.getText()
      val yt = yCoordTF.getText()
      val format = formatChooser.getSelectionModel.getSelectedItem

      for {
        xOffset <- Try(if xt != "" then xt.toInt else 0)
        yOffset <- Try(if yt != "" then yt.toInt else 0)
      } yield FileOpenSettings(StorageCoords(xOffset, yOffset), format)
    }

    val previewPaneBorder = {
      import javafx.scene.layout.*
      import javafx.scene.paint.Color
      val stroke = new BorderStroke(
        Color.RED,
        BorderStrokeStyle.SOLID,
        CornerRadii.EMPTY,
        BorderWidths.DEFAULT
      )
      new Border(stroke)
    }

    val previewPane = {
      val p = new Pane
      p.setMinSize(previewWidth, previewHeight)
      p.setMaxSize(previewWidth, previewHeight)
      p.setBorder(previewPaneBorder)
      p
    }

    val underlyingImage = readImage(previewFile).get

    val wholeImage = new ImageView(SwingFXUtils.toFXImage(underlyingImage.toBufferedImage, null))

    val previewStack = new Pane
    previewStack.getChildren.addAll(wholeImage, previewPane)

    val blankImage = ImageStorage.fill(imageSize, Color.White)
    val imageGridCells = Seq.tabulate(xCount)(x => new GridCell(GridCoords(x, 0), blankImage))
    val imageGrid = ImageGrid.fromCells(imageSize, imageGridCells)

    val (triPreviewPane, updateTriPreview) = ImagePreviewList.fromImageContent(
      imageGrid.images,
      TriImageForPreview.previewSize,
      _ => None
    )

    def updatePreviewAction(): Unit = {
      resultFromInputs() match {
        case Success(FileOpenSettings(StorageCoords(sx, sy), format)) =>
          previewPane.setLayoutX(sx)
          previewPane.setLayoutY(sy)

          for x <- 0 until xCount do {
            val offset = StorageCoords(sx + x * imageSize, sy)
            val newPreviewImage = ImageStorage
              .fromRegularImage(underlyingImage, offset, format, imageSize)
              .getOrElse(blankImage)

            imageGrid.replaceImage(GridCoords(x, 0), newPreviewImage)

            updateTriPreview(_ => ())
          }
        case _ =>
      }
    }

    updatePreviewAction()

    xCoordTF.textProperty.addListener(_ => updatePreviewAction())
    yCoordTF.textProperty.addListener(_ => updatePreviewAction())
    formatChooser.setOnAction(_ => updatePreviewAction())

    val inputForm = makeGridPane(
      Seq(
        Seq(new Label("X coordinate:"), xCoordTF),
        Seq(new Label("Y coordinate:"), yCoordTF),
        Seq(new Label("Format:"), formatChooser)
      )
    )

    getValueFromCustomDialog[FileOpenSettings](
      title = "Open image",
      headerText = "Which part of the image should be opened? Please enter the top left corner:",
      content = Seq(
        inputForm,
        Separator(Orientation.HORIZONTAL),
        previewStack
      ),
      graphic = triPreviewPane,
      resultConverter = {
        case ButtonType.OK => resultFromInputs().getOrElse(null)
        case _             => null
      },
      buttons = Seq(ButtonType.OK, ButtonType.CANCEL)
    )
  }
}

File: ui/src/main/kotlin/tripaint/view/gui/AskForFileOpenSettingsDialog.kt
package tripaint.view.gui

import javafx.collections.FXCollections
import javafx.embed.swing.SwingFXUtils
import javafx.geometry.Orientation
import javafx.scene.control.ButtonType
import javafx.scene.control.ChoiceBox
import javafx.scene.control.Label
import javafx.scene.control.Separator
import javafx.scene.image.ImageView
import javafx.scene.layout.*
import javafx.util.StringConverter
import tripaint.color.Color
import tripaint.coords.GridCoords
import tripaint.coords.StorageCoords
import tripaint.grid.GridCell
import tripaint.grid.ImageGrid
import tripaint.image.ImageStorage
import tripaint.image.RegularImage
import tripaint.image.format.StorageFormat
import tripaint.view.FileOpenSettings
import tripaint.view.JavaFxExt.toFXColor
import tripaint.view.gui.DialogUtils.getValueFromCustomDialog
import tripaint.view.gui.DialogUtils.makeGridPane
import tripaint.view.image.TriImageForPreview
import java.io.File

data class ImagePreviewParams(val previewFile: File, val imageSize: Int, val xCount: Int, val yCount: Int)

object AskForFileOpenSettingsDialog {
    fun askForFileOpenSettings(
        imagePreview: ImagePreviewParams,
        formats: List<Pair<StorageFormat, String>>,
        initiallySelectedFormat: Int,
        readImage: (File) -> RegularImage?
    ): FileOpenSettings? {
        val (previewFile, imageSize, xCount, yCount) = imagePreview
        val (previewWidth, previewHeight) = Pair(xCount * imageSize, yCount * imageSize)

        val xCoordTF = RestrictedTextField.uintTF()
        xCoordTF.promptText = "0"

        val yCoordTF = RestrictedTextField.uintTF()
        yCoordTF.promptText = "0"

        val formatMap: Map<StorageFormat, String> = mapOf(*formats.toTypedArray())

        val formatChooser = run {
            val b = ChoiceBox(FXCollections.observableArrayList(*formats.map { it.first }.toTypedArray()))
            b.selectionModel.select(initiallySelectedFormat)
            b.converter = object : StringConverter<StorageFormat>() {
                override fun toString(t: StorageFormat): String = formatMap[t]!!
                override fun fromString(s: String): StorageFormat = throw UnsupportedOperationException()
            }
            b
        }

        fun resultFromInputs(): Result<FileOpenSettings> {
            val xt = xCoordTF.text
            val yt = yCoordTF.text
            val format = formatChooser.selectionModel.selectedItem

            return Result.runCatching {
                val xOffset = if (xt != "") xt.toInt() else 0
                val yOffset = if (yt != "") yt.toInt() else 0
                FileOpenSettings(StorageCoords.from(xOffset, yOffset), format)
            }
        }

        val previewPaneBorder = run {
            val stroke = BorderStroke(
                    Color.Red.toFXColor(),
            BorderStrokeStyle.SOLID,
            CornerRadii.EMPTY,
            BorderWidths.DEFAULT
            )
            Border(stroke)
        }

        val previewPane = run {
            val p = Pane()
            p.setMinSize(previewWidth.toDouble(), previewHeight.toDouble())
            p.setMaxSize(previewWidth.toDouble(), previewHeight.toDouble())
            p.border = previewPaneBorder
            p
        }

        val underlyingImage = readImage(previewFile)!!

        val wholeImage = ImageView(SwingFXUtils.toFXImage(underlyingImage.toBufferedImage(), null))

        val previewStack = Pane()
        previewStack.children.addAll(wholeImage, previewPane)

        val blankImage = ImageStorage.fill(imageSize, Color.White)
        val imageGridCells = (0..xCount).map { x -> GridCell(GridCoords.from(x, 0), blankImage) }.toList()
        val imageGrid = ImageGrid.fromCells(imageSize, imageGridCells)

        val (triPreviewPane, updateTriPreview) = ImagePreviewList.fromImageContent(
            imageGrid.images(),
            TriImageForPreview.previewSize
        ) { _ -> null }

        fun updatePreviewAction() {
            resultFromInputs().onSuccess {
                val (sc, format) = it
                val sx = sc.x
                val sy = sc.y
                previewPane.setLayoutX(sx.toDouble())
                previewPane.setLayoutY(sy.toDouble())

                for (x in 0 until xCount) {
                    val offset = StorageCoords.from(sx + x * imageSize, sy)
                    val newPreviewImage = ImageStorage
                        .fromRegularImage(underlyingImage, offset, format, imageSize)
                        .getOrDefault(blankImage)

                    imageGrid.replaceImage(GridCoords.from(x, 0), newPreviewImage)

                    updateTriPreview { _ -> }
                }
            }
        }

        updatePreviewAction()

        xCoordTF.textProperty().addListener { _ -> updatePreviewAction() }
        yCoordTF.textProperty().addListener { _ -> updatePreviewAction() }
        formatChooser.setOnAction { _ -> updatePreviewAction() }

        val inputForm = makeGridPane(
            listOf(
                listOf(Label("X coordinate:"), xCoordTF),
                listOf(Label("Y coordinate:"), yCoordTF),
                listOf(Label("Format:"), formatChooser)
            )
        )

        return getValueFromCustomDialog(
            title = "Open image",
            headerText = "Which part of the image should be opened? Please enter the top left corner:",
            content = listOf(
                inputForm,
                Separator(Orientation.HORIZONTAL),
                previewStack
            ),
            graphic = triPreviewPane,
            resultConverter = { r ->
                when (r) {
                    ButtonType.OK -> resultFromInputs().getOrDefault(null)
                    else -> null
                }
            },
            buttons = listOf(ButtonType.OK, ButtonType.CANCEL)
        )
    }
}
